src/java.desktop/share/classes/javax/swing/text/html/CSS.java

```diff
@@ -2604,9 +2672,27 @@ Object toStyleConstants(StyleConstants key, View v) {
             return Float.valueOf(getValue(false));
         }
 
+        @Override
+        public int hashCode() {
+            return Float.hashCode(span);
+        }
+
+        @Override
+        public boolean equals(Object val) {
+            if (percentage) {
+                return val instanceof CSS.LengthValue lu
+                       && Objects.equals(svalue, lu.svalue);
```

Comment:
Doesn't comparing `span` work for percentage values? The comment for `span` field implies it should contain the parsed value.

This comparison could fail for the case where there's a space before the `%` in the string.

Reply:
> This comparison could fail for the case where there's a space before the `%` in the string.

`FontSize` does handle `{"font-size: 100%", "font-size: 100 %"}` pair, but `LengthValue` doesn't.

Reply:
No, it doesn't. For some reason, both `100%` and `200%` are parsed so that `span = 1.0`.

Let's leave it as is then. It handles the most common case.

Handling a space before the percent sign can be postponed to a later fix.

Reply:
Aha, the value is capped at 100%:

https://github.com/openjdk/jdk/blob/4683844d8a26b56c903f6a67aadb159c81c2a2b8/src/java.desktop/share/classes/javax/swing/text/html/CSS.java#L2627-L2628

This is why 200% is parsed as if it were 100%.

The following code
```java
        public boolean equals(Object val) {
            return val instanceof CSS.LengthValue lu
                   && percentage == lu.percentage
                   && span == lu.span
                   && Objects.equals(units, lu.units);
        }
```
works correctly if you modify this line in the test

https://github.com/openjdk/jdk/blob/4683844d8a26b56c903f6a67aadb159c81c2a2b8/test/jdk/javax/swing/text/html/CSS/CSSAttributeEqualityBug.java#L86

to
```java
            {"margin-top: 100%", "margin-top: 50%"},
```

The above code also handles the case `"margin-top: 50 %"` correctly.

Reply:
I had already made that observation in this [comment ](https://github.com/openjdk/jdk/pull/13405#discussion_r1213916093) few days back in case you overlooked
Also, I kept the percentage check for string even though it fails for "space" within string because it seems "space" is not valid in between % value but we can go beyond 100% ie `50 %` is not valid but `200%` is,
 as per https://developer.mozilla.org/en-US/docs/Web/CSS/margin-top
where if you specify `margin-top: 50 %` and then go to other block and come back, you will get a `"X"` but
`margin-top: 200%` is ok

Reply:
> I had already made that observation in this https://github.com/openjdk/jdk/pull/13405#discussion_r1213916093few days back in case you overlooked

I did miss this comment. Sorry about that.

Even though space between the value and the percent sign or the units is invalid (I couldn't find it quickly in the W3C spec for CSS), you should compare the parsed values. We pass `100%` and `200%` but the parsed value is `100%` in both cases — **the attribute sets are indeed *equal***. If you apply either attribute set to a document, you'll get the same result. Does it make sense?

There could be other cases where the computed/parsed values are the same even though the input is different, for example `"font-size: medium"` has a numeric value in points or pixels, so the attribute set with the same value should be equal, don't you agree?

Reply:
I am not sure on 100%, 200% ie >= 100% should be considered equal or not..I could not find in spec...also the URL I gave has 200% as valid value...so as of now I have considered what is normal and made `equals` return false for different percentages irrespective of < or > 100%..

Reply:
I also think that capping at 100% is wrong but it's another bug.

Currently, both 100% and 200% result in attribute sets which behave as if both were 100%, therefore they should be equal. I strongly believe we should compare the parsed values not the input string.

Reply:
OK. Fair enough..guess CSS implementation and equality should match...Modified...
Upper capping and equality can be looked at later if needed